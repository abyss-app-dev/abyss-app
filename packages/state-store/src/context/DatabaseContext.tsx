import type { SQliteClient } from '@abyss/records';
import { createContext, type ReactNode, useContext } from 'react';

interface DatabaseContextType {
    database: SQliteClient | null;
}

const DatabaseContext = createContext<DatabaseContextType>({ database: null });

interface DatabaseProviderProps {
    database: SQliteClient;
    children: ReactNode;
}

export function DatabaseProvider({ database, children }: DatabaseProviderProps) {
    return <DatabaseContext.Provider value={{ database }}>{children}</DatabaseContext.Provider>;
}

export function useDatabase(): SQliteClient {
    const context = useContext(DatabaseContext);
    if (!context.database) {
        throw new Error('useDatabase must be used within a DatabaseProvider');
    }
    return context.database;
}
