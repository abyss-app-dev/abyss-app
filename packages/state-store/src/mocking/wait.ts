export function wait(ms = 50) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
