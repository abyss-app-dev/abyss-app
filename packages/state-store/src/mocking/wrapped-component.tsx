import type { StateRecorder } from './state-recorder';

export function TestRecorder<T>({ functionToTest, recorder }: { functionToTest: () => T; recorder: StateRecorder }) {
    recorder.record(functionToTest());
    return <>content</>;
}
