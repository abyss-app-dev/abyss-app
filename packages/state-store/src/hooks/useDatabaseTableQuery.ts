import type { SqliteTables } from '@abyss/records';
import { useCallback, useEffect, useRef } from 'react';
import { useDatabase } from '../context/DatabaseContext';
import { useQuery } from './useQuery';

export type DatabaseTableQueryFunction<T extends keyof SqliteTables, R> = (table: SqliteTables[T]) => R | Promise<R>;

export function useDatabaseTableQuery<T extends keyof SqliteTables, R>(table: T, queryFunction: DatabaseTableQueryFunction<T, R>) {
    const database = useDatabase();

    // Use ref to store the latest query function to avoid dependency issues
    const queryFunctionRef = useRef(queryFunction);
    queryFunctionRef.current = queryFunction;

    // Stabilize the query function to prevent infinite re-renders
    const stableQueryFunction = useCallback((tableInstance: SqliteTables[T]) => queryFunctionRef.current(tableInstance), []);

    const query = useQuery(async () => {
        return stableQueryFunction(database.tables[table]);
    }, [table, stableQueryFunction]);

    useEffect(() => {
        return database.tables[table].subscribe(async () => {
            query.refetch();
        });
    }, [database, table, query.refetch]);

    return query;
}
