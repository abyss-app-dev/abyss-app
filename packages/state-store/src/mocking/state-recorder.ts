import { render } from '@testing-library/react';
import type { ReactNode } from 'react';
import { expect } from 'vitest';
import { wait } from './wait';

export class StateRecorder {
    private states: unknown[] = [];

    record(state: unknown): void {
        this.states.push(state);
    }

    async delayedResolve(): Promise<unknown[]> {
        await wait();
        return this.states;
    }

    getStates(): unknown[] {
        return this.states;
    }

    getLatestState(): unknown {
        return this.states[this.states.length - 1];
    }

    clear(): void {
        this.states.length = 0;
    }

    async renderToSnapshot(component: ReactNode): Promise<void> {
        render(component);
        expect(await this.delayedResolve()).toMatchSnapshot();
    }
}
