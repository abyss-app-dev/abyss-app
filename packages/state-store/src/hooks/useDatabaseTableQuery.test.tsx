import { buildTestDB, type ReferencedSettingsTable, type SQliteClient, SqliteTable } from '@abyss/records';
import { useEffect, useState } from 'react';
import { beforeEach, describe, test } from 'vitest';
import { DatabaseProvider } from '../context/DatabaseContext';
import { StateRecorder } from '../mocking/state-recorder';
import { TestRecorder } from '../mocking/wrapped-component';
import { useDatabaseTableQuery } from './useDatabaseTableQuery';

describe('useDatabaseTableQuery hook', () => {
    let db: SQliteClient;
    let recorder: StateRecorder;

    beforeEach(async () => {
        db = await buildTestDB();
        recorder = new StateRecorder();
    });

    test('applies query function to table', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseTableQuery(SqliteTable.settings, table => `Table: ${table.constructor.name}`);
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('async query function on table', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseTableQuery(SqliteTable.settings, async table => {
                            await new Promise(resolve => setTimeout(resolve, 10));
                            return `Async table: ${table.constructor.name}`;
                        });
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('complex table query transformation', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const result = useDatabaseTableQuery(SqliteTable.settings, table => {
                            return {
                                tableName: table.constructor.name,
                                hasDefaultMethod: typeof table.default === 'function',
                                hasGetMethod: typeof table.get === 'function',
                                tableInfo: 'settings-table-metadata',
                            };
                        });
                        return result;
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('query function changes dynamically', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [version, setVersion] = useState(1);

                        const queryFn =
                            version === 1
                                ? (table: ReferencedSettingsTable) => `v1-${table.constructor.name}`
                                : (table: ReferencedSettingsTable) => `v2-meta-${table.constructor.name}`;

                        const result = useDatabaseTableQuery(SqliteTable.settings, queryFn);

                        useEffect(() => {
                            if (version === 1) {
                                setTimeout(() => setVersion(2));
                            }
                        }, [version]);

                        return { ...result, version };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('manual refetch functionality', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasRefetched, setHasRefetched] = useState(false);

                        const result = useDatabaseTableQuery(SqliteTable.settings, table => `Refetch test: ${table.constructor.name}`);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasRefetched) {
                                setHasRefetched(true);
                                setTimeout(() => result.refetch());
                            }
                        }, [result.data, result.loading, result.refetch, hasRefetched]);

                        return { ...result, hasRefetched };
                    }}
                />
            </DatabaseProvider>
        );
    });

    test('subscription updates with table processing', async () => {
        await recorder.renderToSnapshot(
            <DatabaseProvider database={db}>
                <TestRecorder
                    recorder={recorder}
                    functionToTest={() => {
                        const [hasSetData, setHasSetData] = useState(false);

                        const result = useDatabaseTableQuery(SqliteTable.settings, table => `Processed: ${table.constructor.name}`);

                        useEffect(() => {
                            if (result.data && !result.loading && !hasSetData && result.setData) {
                                setHasSetData(true);
                                setTimeout(() => result.setData('manually-set-table-data'));
                            }
                        }, [result.data, result.loading, result.setData, hasSetData]);

                        return { ...result, hasSetData };
                    }}
                />
            </DatabaseProvider>
        );
    });
});
