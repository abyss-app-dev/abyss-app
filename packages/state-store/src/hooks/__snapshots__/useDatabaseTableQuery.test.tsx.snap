// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useDatabaseTableQuery hook > applies query function to table 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Table: _ReferencedSettingsTable",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseTableQuery hook > async query function on table 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Async table: _ReferencedSettingsTable",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseTableQuery hook > complex table query transformation 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": {
      "hasDefaultMethod": true,
      "hasGetMethod": true,
      "tableInfo": "settings-table-metadata",
      "tableName": "_ReferencedSettingsTable",
    },
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseTableQuery hook > manual refetch functionality 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasRefetched": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: _ReferencedSettingsTable",
    "error": null,
    "hasRefetched": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: _ReferencedSettingsTable",
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Refetch test: _ReferencedSettingsTable",
    "error": null,
    "hasRefetched": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;

exports[`useDatabaseTableQuery hook > query function changes dynamically 1`] = `
[
  {
    "data": null,
    "error": null,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
    "version": 1,
  },
  {
    "data": "v1-_ReferencedSettingsTable",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "version": 1,
  },
  {
    "data": "v1-_ReferencedSettingsTable",
    "error": null,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
    "version": 2,
  },
]
`;

exports[`useDatabaseTableQuery hook > subscription updates with table processing 1`] = `
[
  {
    "data": null,
    "error": null,
    "hasSetData": false,
    "loading": true,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Processed: _ReferencedSettingsTable",
    "error": null,
    "hasSetData": false,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "Processed: _ReferencedSettingsTable",
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
  {
    "data": "manually-set-table-data",
    "error": null,
    "hasSetData": true,
    "loading": false,
    "refetch": [Function],
    "setData": [Function],
  },
]
`;
